### PR TITLE
Add markus.utils.generate_tag

### PR DESCRIPTION
**Pull request by willkg**
Adds a utility function for sanitizing keys and values for tags.

### CODE REVIEW

**Comment by willkg** [comment]
@osmose Can you take a look at this? Does it look ok?

File: docs/metricsoverview.rst [hunk 1]
```diff
@@ -101,3 +101,8 @@ Generally, if you follow these rules, your tags should work across all backends:
    tag names.
 2. Tag names should begin with a letter.
 3. Tag names should be short.
+4. Tag value domains should be finite and small-ish like under 1,000. Be wary of
+   tags around ip addresses, usernames,
```
Comment (Osmose) [inline docs/metricsoverview.rst]: Unfinished sentence?

File: docs/metricsoverview.rst [hunk 2]
```diff
@@ -101,3 +101,8 @@ Generally, if you follow these rules, your tags should work across all backends:
    tag names.
 2. Tag names should begin with a letter.
 3. Tag names should be short.
+4. Tag value domains should be finite and small-ish like under 1,000. Be wary of
```
Comment (Osmose) [inline docs/metricsoverview.rst]: "Tag value domains" is making this kinda hard for me to read. Do you mean that you should only have like under 1000 different tag values?

Maybe "Tag values should be limited to a small (under 1000) set of possible values."?

File: markus/utils.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if not isinstance(value, six.string_types + (NONE_TYPE,)):
+        raise ValueError('value must be None or a string type, but got %r instead' % value)
+
+    # Sanitize the key
+    key = BAD_TAG_CHAR_REGEXP.sub('_', key).strip()
+
+    # Build the tag
+    if value is None or not value.strip():
+        tag = key
+
```
Comment (Osmose) [inline markus/utils.py]: I did not expect a newline here.

File: docs/metricsoverview.rst (hunk 1 above)
Comment (willkg) [reply docs/metricsoverview.rst]: Oops--thanks!

File: docs/metricsoverview.rst (hunk 2 above)
Comment (willkg) [reply docs/metricsoverview.rst]: I like that wording a lot better--thank you!

**Comment by willkg** [comment]
> Would it be easier if Markus auto-sanitized incoming tags instead of just providing a utility function for them?

I talk about this a bit in https://github.com/willkg/markus/issues/25#issuecomment-343969910 .

Thank you for looking through this!

**Pull request merged**